File: app/src/main/java/com/example/vn008xw/carbeat/ui/movie/MovieViewModel.java
package com.example.vn008xw.carbeat.ui.movie;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.Transformations;
import android.arch.lifecycle.ViewModel;
import android.support.annotation.NonNull;
import android.support.annotation.VisibleForTesting;
import android.util.Pair;

import com.example.vn008xw.carbeat.data.repository.FavoriteMovieRepository;
import com.example.vn008xw.carbeat.data.repository.ImageRepository;
import com.example.vn008xw.carbeat.data.repository.MovieRepository;
import com.example.vn008xw.carbeat.data.vo.AbsentLiveData;
import com.example.vn008xw.carbeat.data.vo.FavoriteMovie;
import com.example.vn008xw.carbeat.data.vo.ImageResult;
import com.example.vn008xw.carbeat.data.vo.Movie;
import com.example.vn008xw.carbeat.data.vo.Resource;

import javax.inject.Inject;

/**
 * Created by vn008xw on 7/11/17.
 */

public class MovieViewModel extends ViewModel {

  @VisibleForTesting
  final MovieRepository movieRepository;
  @VisibleForTesting
  final ImageRepository imageRepository;
  @VisibleForTesting
  final FavoriteMovieRepository favoriteRepository;
  @VisibleForTesting
  final MutableLiveData<Integer> movieId;
  @VisibleForTesting
  final LiveData<Resource<Movie>> movieResult;
  @VisibleForTesting
  final LiveData<Resource<FavoriteMovie>> favoriteMovie;
  @VisibleForTesting
  final LiveData<Resource<ImageResult>> images;
  @VisibleForTesting
  final MutableLiveData<Pair<Integer, Boolean>> saved;

  @SuppressWarnings("unchecked")
  @Inject
  public MovieViewModel(MovieRepository movieRepository,
                        ImageRepository imageRepository,
                        FavoriteMovieRepository favoriteMovieRepository) {
    this.movieRepository = movieRepository;
    this.imageRepository = imageRepository;
    favoriteRepository = favoriteMovieRepository;

    movieId = new MutableLiveData<>();
    saved = new MutableLiveData<>();

    movieResult = Transformations.switchMap(movieId, id->{
      if (id == null) return AbsentLiveData.create();
      return movieRepository.loadMovie(id);
    });

    favoriteMovie = Transformations.switchMap(saved, pair->{
      if (pair == null) return AbsentLiveData.create();
      return favoriteMovieRepository.loadFavoriteMovie(pair.first);
    });

    images = Transformations.switchMap(movieId, id->{
      if (id == null) return AbsentLiveData.create();
      return imageRepository.loadImages(id);
    });
  }

  void setFind(@NonNull Integer id, @NonNull Boolean find) {
    saved.setValue(Pair.create(id, find));
  }

  void setMovieId(@NonNull Integer id) {
    movieId.setValue(id);
  }

  LiveData<Resource<Movie>> getMovie() {
    return movieResult;
  }

  LiveData<Resource<ImageResult>> getImages() {
    return images;
  }

  LiveData<Resource<FavoriteMovie>> getFavoriteMovie() {
    return favoriteMovie;
  }

  void saveMovie(@NonNull Movie movie) {
    final FavoriteMovie favoriteMovie = new FavoriteMovie(movie.getTitle(), movie);
    favoriteRepository.saveFavoriteMovie(favoriteMovie);
    saved.setValue(Pair.create(movie.getId(), true));
  }
}

File: app/src/main/java/com/example/vn008xw/carbeat/ui/movie/MovieViewModel.kt
package com.example.vn008xw.carbeat.ui.movie

import android.arch.lifecycle.*
import android.support.annotation.VisibleForTesting
import android.util.Log
import com.example.vn008xw.carbeat.data.repository.FavoriteMovieRepository
import com.example.vn008xw.carbeat.data.repository.ImageRepository
import com.example.vn008xw.carbeat.data.repository.MovieRepository
import com.example.vn008xw.carbeat.data.vo.*
import javax.inject.Inject


class MovieViewModel @Inject
constructor(@VisibleForTesting
            internal val movieRepository: MovieRepository,
            @VisibleForTesting
            internal val imageRepository: ImageRepository,
            @VisibleForTesting
            val favoriteMovieRepository: FavoriteMovieRepository) : ViewModel() {

  @VisibleForTesting
  internal val movieId = MutableLiveData<Int>()
  @VisibleForTesting
  internal val movieResult: LiveData<Resource<Movie>>
  @VisibleForTesting
  internal val images: LiveData<Resource<ImageResult>>
  @VisibleForTesting
  internal val favoriteMovie: LiveData<Resource<FavoriteMovie>>
  @VisibleForTesting
  internal val saved = MutableLiveData<Pair<Int, Boolean>>()

  init {
    movieResult = Transformations.switchMap(movieId) { id ->
      if (id == null) AbsentLiveData.create<Resource<Movie>>()
      else {
        movieRepository.loadMovie(id)
      }
    }
    images = Transformations.switchMap(movieId) { id ->
      if (id == null) AbsentLiveData.create<Resource<ImageResult>>()
      else {
        imageRepository.loadImages(id)
      }
    }
    favoriteMovie = Transformations.switchMap(saved) { id ->
      if (id == null) AbsentLiveData.create<Resource<FavoriteMovie>>()
      else {
        favoriteMovieRepository.loadFavoriteMovie(id.first)
      }
    }
  }

  fun setFind(id: Int, find: Boolean) {
    saved.value = Pair(id, find)
  }

  fun setMovieId(id: Int) {
    movieId.value = id
  }

  fun getMovie() = movieResult

  fun getImages() = images

  fun getFavoriteMovie() = favoriteMovie

  fun saveMovie(movie: Movie) {
    val favoriteMovie = FavoriteMovie(movie.title, movie)
    favoriteMovieRepository.saveFavoriteMovie(favoriteMovie)
    saved.value = Pair(movie.id, true)
  }
}